lib/Transforms/ShapeInference/ShapeInference.cpp

```diff
@@ -0,0 +1,187 @@
+  InferShape(MLIRContext *context)
+      : OpRewritePattern<func::FuncOp>(context, /*benefit=*/1) {}
+
+ public:
+  LogicalResult matchAndRewrite(func::FuncOp funcOp,
+                                PatternRewriter &rewriter) const override {
+    /////////////////////////////////////////////////////////////////////
+    //// Step 1: Check for shape-annotated dynamic-sized tensor args ////
+    /////////////////////////////////////////////////////////////////////
+
+    // find all non-static-shape ranked tensor types in the function's arguments
+    std::vector<std::tuple<int, RankedTensorType>> dynTensorArgs = {};
+    for (auto [idx, type] :
+         llvm::enumerate(funcOp.getFunctionType().getInputs())) {
+      auto rankedType = dyn_cast<RankedTensorType>(type);
+      if (!rankedType || rankedType.hasStaticShape()) continue;
+      dynTensorArgs.emplace_back(idx, rankedType);
+    }
+    // no non-static-shape tensor types -> exit
+    if (dynTensorArgs.empty()) return failure();
+
+    // Check if there are even argument attributes at all.
+    // If not, there is nothing to infer from -> exit
+    auto optionalAttrs = funcOp.getArgAttrs();
+    if (!optionalAttrs) return failure();
+    auto attrs = optionalAttrs->getValue();
+
+    // Update types with annotations (where it exists)
+    bool changed = false;
+    for (auto &[idx, type] : dynTensorArgs) {
+      // Check if the current argument has any attributes.
+      auto dictattr = dyn_cast<DictionaryAttr>(attrs[idx]);
+      if (!dictattr) continue;  // nothing to infer from
+
+      // Get the shape attribute if it exists
+      auto *shapeAttr = llvm::find_if(dictattr, [&](NamedAttribute attr) {
+        return attr.getName() == "shape.shape";
+      });
+      if (!shapeAttr) continue;  // nothing to infer from
+
+      // Extract the actual shape from the shape annotation attribute
+      auto arrayAttr = dyn_cast<ArrayAttr>(shapeAttr->getValue());
+      if (!arrayAttr)
+        return emitError(funcOp.getLoc(),
+                         "Invalid shape attribute (expected ArrayAttr)");
+
+      if (llvm::any_of(arrayAttr,
+                       [](Attribute attr) { return !isa<IntegerAttr>(attr); }))
+        return emitError(
+            funcOp.getLoc(),
+            "Invalid shape attribute (expected ArrayAttr of only IntegerAttr)");
+
+      SmallVector<int64_t> shape = llvm::to_vector(
+          llvm::map_range(arrayAttr.getAsRange<IntegerAttr>(),
+                          [](IntegerAttr i) { return i.getInt(); }));
+
+      if (shape.size() != type.getRank())
+        return emitError(
+            funcOp.getLoc(),
+            llvm::formatv("Shape annotation has wrong number of elements "
+                          "(expected {0}, got {1}).",
+                          type.getRank(), shape.size()));
+
+      // save the updated argument type
+      type = RankedTensorType::get(shape, type.getElementType());
```

Comment:
This actually mutates the container being looped over?? :fearful:

Reply:
It does, but not in a way that would invalidate the iterators (afaik). I'm not sure there's a much nicer way to express this except the trivial approach of taking a copy of the container beforehand?

Reply:
It's not the end of the world, but IMO a better design would be to split this step off into a helper function that returns or populates a new vector containing the materialized types, or else just merge the loop from 31-36 with this loop so that the emplace_back on line 35 can be done here instead (the check on func attrs can be moved to the beginning, and the empty check on line 38 is effectively a duplicate of like 87, paired with the continue on line 34 that would be merged into this loop)